lib/puppet/indirector/catalog/compiler.rb

```diff
@@ -128,7 +128,33 @@ def inline_metadata(catalog, checksum_type)
       file = resource.to_ral
 
       if file.recurse?
-        # TODO: find and replace metadata for children recursively
+        file.recurse_remote_metadata.each do |meta|
+          # Don't create a new resource for the parent directory
+          next if meta.relative_path == "."
+
+          # TODO: Conditionally copy owner, group, and mode when we can check if permissions are set
```

Comment:
Lines 137-143 seem like they could be combined with replace_metadata pretty reasonably. That seems like it'd make ensuring the correct metadata is updated easier.